File: kt/aoc2023-lib/src/commonMain/kotlin/com/github/ephemient/aoc2023/Day11.kt
package com.github.ephemient.aoc2023

class Day11(input: String) {
    private val xs: List<Int>
    private val ys: List<Int> = buildList {
        xs = input.lines().map { line ->
            var count = 0
            for ((i, char) in line.withIndex()) {
                if (char == '#') {
                    while (lastIndex < i) add(0)
                    this[i]++
                    count++
                }
            }
            count
        }
    }

    fun part1(): Long = solve(2)

    fun part2(): Long = solve(1000000)

    internal fun solve(n: Long): Long = solve1(xs, n) + solve1(ys, n)

    private fun solve1(data: List<Int>, n: Long): Long {
        var total = 0L
        for ((i, a) in data.withIndex()) {
            if (a == 0) continue
            var m = 0L
            for (j in i + 1..data.lastIndex) {
                val b = data[j]
                m += if (b == 0) n else 1
                total += m * a * b
            }
        }
        return total
    }
}

File: rs/src/day11.rs
fn solve(data: &str, n: usize) -> usize {
    let mut counts = Vec::<usize>::new();
    solve1(
        &data
            .lines()
            .map(|line| {
                let mut count = 0;
                counts.extend(std::iter::repeat(0).take(line.len()).skip(counts.len()));
                for (i, c) in line.char_indices() {
                    if c == '#' {
                        counts[i] += 1;
                        count += 1;
                    }
                }
                count
            })
            .collect::<Vec<_>>(),
        n,
    ) + solve1(&counts, n)
}

fn solve1(data: &[usize], n: usize) -> usize {
    data.iter()
        .enumerate()
        .map(|(i, &a)| {
            if a == 0 {
                0
            } else {
                data[i + 1..]
                    .iter()
                    .scan(0, move |m, &b| {
                        *m += if b == 0 { n } else { 1 };
                        Some(*m * a * b)
                    })
                    .sum()
            }
        })
        .sum()
}

pub fn part1(data: &str) -> usize {
    solve(data, 2)
}

pub fn part2(data: &str) -> usize {
    solve(data, 1000000)
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &str = indoc! {"
        ...#......
        .......#..
        #.........
        ..........
        ......#...
        .#........
        .........#
        ..........
        .......#..
        #...#.....
    "};

    #[test]
    fn part1_examples() {
        assert_eq!(374, solve(EXAMPLE, 2));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(1030, solve(EXAMPLE, 10));
        assert_eq!(8410, solve(EXAMPLE, 100));
    }
}
